File: back/src/main/java/com/back/global/security/oauth2/loginhandler/OAuth2LoginFailureHandler.java
package com.back.global.security.oauth2.loginhandler;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.stereotype.Component;
import org.springframework.web.util.UriComponentsBuilder;

import java.io.IOException;

@Component
@RequiredArgsConstructor
public class OAuth2LoginFailureHandler implements AuthenticationFailureHandler {
    private final OAuth2RedirectHandler redirectHandler;

    @Override
    public void onAuthenticationFailure(
            HttpServletRequest request,
            HttpServletResponse response,
            AuthenticationException exception
    ) throws IOException, ServletException {
        String errorCode = "oauth2_login_failed";

        if (exception instanceof OAuth2AuthenticationException oauth2Exception) {
            errorCode = oauth2Exception.getError().getErrorCode();
        }

        redirectHandler.redirectFailure(response, errorCode);
    }
}

File: back/src/main/kotlin/com/back/global/security/oauth2/loginhandler/OAuth2LoginFailureHandler.kt
package com.back.global.security.oauth2.loginhandler

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import com.back.domain.auth.repository.SocialLinkCookieRepository
import org.springframework.security.core.AuthenticationException
import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.stereotype.Component

@Component
class OAuth2LoginFailureHandler(
    private val redirectHandler: OAuth2RedirectHandler,
    private val socialLinkCookieRepository: SocialLinkCookieRepository,
) : AuthenticationFailureHandler {

    override fun onAuthenticationFailure(
        request: HttpServletRequest,
        response: HttpServletResponse,
        exception: AuthenticationException
    ) {
        val isSocialLink = socialLinkCookieRepository.load() != null
        socialLinkCookieRepository.remove()
        val errorCode = (exception as? OAuth2AuthenticationException)
            ?.error
            ?.errorCode
            ?: "oauth2_login_failed"

        if (isSocialLink) {
            redirectHandler.redirectLinkFailure(response, errorCode)
        } else {
            redirectHandler.redirectLoginFailure(response, errorCode)
        }
    }
}
